hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/ContainerReportQueue.java

```diff
@@ -0,0 +1,399 @@
+    return report instanceof IncrementalContainerReportFromDatanode;
+  }
+
+  private VALUE removeAndGet(String uuid) {
+    if (uuid == null) {
+      return null;
+    }
+
+    List<VALUE> dataList = dataMap.get(uuid);
+    VALUE report = null;
+    if (dataList != null && !dataList.isEmpty()) {
+      report = dataList.remove(0);
+      --capacity;
+      if (dataList.isEmpty()) {
+        dataMap.remove(uuid);
+      }
+    }
+    return report;
+  }
+
+  private VALUE getReport(String uuid) {
+    if (uuid == null) {
+      return null;
+    }
+
+    List<VALUE> dataList = dataMap.get(uuid);
+    if (dataList != null && !dataList.isEmpty()) {
+      return dataList.get(0);
+    }
+    return null;
+  }
+
+  private static void checkNotNull(Object v) {
+    if (v == null) {
+      throw new NullPointerException();
+    }
+  }
+
+  public boolean addValue(@NotNull VALUE value) {
+    synchronized (this) {
+      if (remainingCapacity() == 0) {
+        return false;
+      }
+
+      if (isFCRReport(value)) {
+        return addContainerReport(value);
+      } else if (isICRReport(value)) {
+        return addIncrementalReport(value);
+      }
+      return false;
+    }
+  }
+
+  @Override
+  public boolean add(@NotNull VALUE value) {
+    checkNotNull(value);
+    synchronized (this) {
+      if (remainingCapacity() == 0) {
+        throw new IllegalStateException("capacity not available");
+      }
+
+      return addValue(value);
+    }
+  }
+
+  @Override
+  public boolean offer(@NotNull VALUE value) {
+    checkNotNull(value);
+    synchronized (this) {
+      return addValue(value);
+    }
+  }
+
+  @Override
+  public VALUE remove() {
+    synchronized (this) {
+      String uuid = orderingQueue.remove();
+      if (null == uuid) {
+        throw new NoSuchElementException();
+      }
+      return removeAndGet(uuid);
+    }
+  }
+
+  @Override
+  public VALUE poll() {
+    synchronized (this) {
+      String uuid = orderingQueue.poll();
+      return removeAndGet(uuid);
+    }
+  }
+
+  @Override
+  public VALUE element() {
+    synchronized (this) {
+      String uuid = orderingQueue.element();
+      if (null == uuid) {
+        throw new NoSuchElementException();
+      }
+      return getReport(uuid);
+    }
+  }
+
+  @Override
+  public VALUE peek() {
+    synchronized (this) {
+      String uuid = orderingQueue.peek();
+      return getReport(uuid);
+    }
+  }
+
+  @Override
+  public void put(@NotNull VALUE value) throws InterruptedException {
+    checkNotNull(value);
+    while (!addValue(value)) {
+      Thread.currentThread().sleep(10);
+    }
+  }
+
+  @Override
+  public boolean offer(VALUE value, long timeout, @NotNull TimeUnit unit)
+      throws InterruptedException {
+    checkNotNull(value);
+    long timeoutMillis = unit.toMillis(timeout);
+    while (timeoutMillis > 0) {
+      if (addValue(value)) {
+        return true;
+      }
+      Thread.currentThread().sleep(10);
+      timeoutMillis -= 10;
```

Comment:
Should it deduct actual time difference instead of assuming exactly 10 milliseconds has passed?

Reply:
used startTime and end time to get actual difference and updated same.